src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+            uint64_t offsetInCPUBuffer = mainRegion.bufferOffset + core::dot(localImageOffset, srcBufferByteStrides)[0];
+            uint8_t* inCpuBufferPointer = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(srcBuffer->getPointer()) + offsetInCPUBuffer);
+            asset::ICPUImage::SCreationParams inCPUImageParams = dstImageParams;
+            inCPUImageParams.flags = asset::IImage::ECF_NONE; // Because we may want to write to first few layers of CUBEMAP (<6) but it's not valid to create an Cube ICPUImage with less that 6 layers.
+            inCPUImageParams.format = srcImageFormat;
+            inCPUImageParams.extent = regionToCopyNext.imageExtent;
+            inCPUImageParams.arrayLayers = regionToCopyNext.imageSubresource.layerCount;
+            inCPUImageParams.mipLevels = 1u;
+            inCPUImage = asset::ICPUImage::create(std::move(inCPUImageParams));
+            assert(inCPUImage);
+            core::smart_refctd_ptr<asset::ICPUBuffer> inCPUBuffer = core::make_smart_refctd_ptr< asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>, true> >(srcBuffer->getSize(), inCpuBufferPointer, core::adopt_memory);
+            inCPUImage->setBufferAndRegions(std::move(inCPUBuffer), inCpuImageRegionsDynArray);
+            assert(inCPUImage->getBuffer());
+            assert(inCPUImage->getRegions().size() > 0u);
+        }
+
+        // outCPUImage is an image matching the params of dstImage but with the extents and layer count of the current region being copied and mipLevel 1u
+        // the buffer of this image is set to (stagingBufferPointer + stagingBufferOffset) and the related region is set to cover the whole copy region (offset from 0)
+        {
+            auto outCpuImageRegionsDynArray = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1);
+            auto& outCpuImageRegion = outCpuImageRegionsDynArray->front();
+            outCpuImageRegion = {};
+            outCpuImageRegion.bufferOffset = 0u;
+            outCpuImageRegion.bufferRowLength = regionToCopyNext.bufferRowLength;
+            outCpuImageRegion.bufferImageHeight = regionToCopyNext.bufferImageHeight;
+            outCpuImageRegion.imageSubresource.aspectMask = mainRegion.imageSubresource.aspectMask;
+            outCpuImageRegion.imageSubresource.mipLevel = 0u;
+            outCpuImageRegion.imageSubresource.baseArrayLayer = 0u;
+            outCpuImageRegion.imageOffset.x = 0u;
+            outCpuImageRegion.imageOffset.y = 0u;
+            outCpuImageRegion.imageOffset.z = 0u;
+            outCpuImageRegion.imageExtent.width    = regionToCopyNext.imageExtent.width;
+            outCpuImageRegion.imageExtent.height   = regionToCopyNext.imageExtent.height;
+            outCpuImageRegion.imageExtent.depth    = regionToCopyNext.imageExtent.depth;
+            outCpuImageRegion.imageSubresource.layerCount = core::max(regionToCopyNext.imageSubresource.layerCount, 1u);
```

Comment:
this arithmetic would get much simpler if the two images were made up-front and were persistent.